drivers: can: add CAN transceiver support

henrikbrixandersen opened the pull request:
Add support for CAN transceivers:
- Added new devicetree bindings for CAN transceivers. These bindings are in-line with how CAN transceivers are represented in upstream Linux DTS.
- Added CAN transceivers to all relevant in-tree board devicetrees.
- Added new device driver API for CAN transceivers.
- Added generic GPIO-controlled CAN transceiver with support for a combination of enable and standby GPIOs.
- Updated all in-tree CAN drivers with CAN transceiver support.
- Added API support for determining the maximum bitrate supported by a CAN controller and attached CAN transceiver.
- Guard against setting a too high bitrate.

henrikbrixandersen committed:
dts: bindings: add devicetree bindings for CAN transceivers

Add generic devicetree bindings for simple CAN transceivers.

Always-on CAN transceivers are considered passive and just provide a
maximum supported bitrate.

Active CAN controllers can typically be controlled by the MCU via either
SPI, I2C, or GPIO. Common GPIO controlled CAN transceivers provide
either a stand-by or an enable pin (or both) for controlling the state
of the CAN transceiver.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen committed:
boards: arm: lpcxpresso55s16: add CAN transceiver

The NXP LPCXpresso55S16 development board is equipped with a NXP TJA1044
CAN transceiver with a maximum bitrate of 5Mbit/s.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen committed:
boards: arm: mimxrt1xxx: add CAN transceiver

The NXP MIMXRT10xx/11xx EVK boards are equipped with NXP TJA1057 CAN
transceivers with a maximum bitrate of 5Mbit/s.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen committed:
boards: arm: olimex_stm32_p405: add CAN transceiver

The Olimex STM32-P405 development board is equipped with a TI SN65HVD230
CAN transceiver with a maximum bitrate of 1Mbit/s.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen committed:
boards: arm: olimexino_stm32: add CAN transceiver

The Olimex OLIMEXINO-STM32 development board is equipped with a
Microchip MCP2551 CAN transceiver with a maximum bitrate of 1Mbit/s.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen committed:
boards: arm: rcar_h3ulcb_cr7: add CAN transceiver

The Renesas R-Car H3ULCB development board is equipped with a TI
TCAN332G CAN transceiver with a maximum bitrate of 5Mbit/s.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen committed:
boards: arm: rddrone_fmuk66: add CAN transceiver

The NXP RDDRONE-FMUK66 development board is equipped with dual NXP
TJA1042 CAN transceivers with a maximum bitrate of 5Mbit/s.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen committed:
boards: arm: sam_e70_xplained: add CAN transceiver

The Atmel SAM E70(B) Xplained development board is equipped with a
Microchip ATA6561 CAN transceiver with a maximum bitrate of 5Mbit/s.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen committed:
boards: arm: twr_ke18f: add CAN transceiver

The NXP TWR-KE18F development board is equipped with a NXP MC33901 CAN
transceiver with a maximum bitrate of 1Mbit/s.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen committed:
boards: shields: dfrobot_can_bus_v2_0: add CAN transceiver

The DFRobot CAN BUS Shield V2.0 is equipped with a NXP TJA1050 CAN
transceiver with a maximum bitrate of 1Mbit/s.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen committed:
boards: shields: keyestudio_can_bus_ks0411: add CAN transceiver

The Keyestudio CAN-BUS Shield (KS0411) is equipped with a Microchip
MCP2551 CAN transceiver with a maximum bitrate of 1Mbit/s.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen committed:
drivers: can: add generic GPIO controlled CAN transceiver driver

Add generic driver for GPIO controlled CAN transceivers.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen committed:
drivers: can: add CAN transceiver device driver API

Add API for controlling the state of a CAN transceiver.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen committed:
drivers: can: add support for getting the maximum supported bitrate

Add support for getting the maximum supported bitrate in bits/s for CAN
controller/transceiver combination and check that a requested bitrate is
within the supported range.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen committed:
drivers: can: mcux: flexcan: add transceiver support

Add support for CAN transceivers to the NXP MCUX FlexCAN driver.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen committed:
drivers: can: stm32: add transceiver support

Add support for CAN transceivers to the ST STM32 bxCAN driver.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen committed:
drivers: can: rcar: add transceiver support

Add support for CAN transceivers to the Renesas R-Car CAN driver.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen committed:
drivers: can: mcp2515: add transceiver support

Add support for CAN transceivers to the Microchip MCP2515 CAN driver.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen committed:
drivers: can: mcan: add transceiver support

Add support for CAN transceivers to the Bosch M_CAN CAN drivers.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

martinjaeger commented on drivers/can/can_handlers.c:
```diff
@@ -30,6 +30,17 @@ static inline int z_vrfy_can_get_core_clock(const struct device *dev,
+static inline int z_vrfy_can_get_max_bitrate(const struct device *dev,
+					     uint32_t *max_bitrate)
+{
+
```
Nits: Unnecessary empty line.

henrikbrixandersen replied:
Fixed.

martinjaeger commented on dts/bindings/phy/can-transceiver.yaml:
```diff
@@ -0,0 +1,16 @@
+
+# Common fields for CAN transceivers
+
+include: phy-controller.yaml
```
This is not existing, as far as I can see.

henrikbrixandersen replied:
https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/bindings/phy/phy-controller.yaml ?

martinjaeger replied:
Sorry, never mind. For some reason didn't find it during my search through the Zephyr tree.

martinjaeger reviewed (commented):
Great to get automatic handling of the CAN transceiver standby pin. Thanks for the PR!

I'm not completely finished with the review, yet. Here are two minor comments/questions already.

henrikbrixandersen commented:
Rebased to resolve `CODEOWNERS` merge conflict.

aaillet commented on drivers/can/can_rcar.c:
```diff
@@ -1056,6 +1092,9 @@ static const struct can_driver_api can_rcar_driver_api = {
 		.pcfg = PINCTRL_DT_INST_DEV_CONFIG_GET(n),			\
+		.phy = DEVICE_DT_GET_OR_NULL(DT_INST_PHANDLE(n, phys)),		\
+		.max_bitrate =							\
+			CAN_TRANSCEIVER_DT_MAX_BITRATE_INST_GET(n, 1000000),	\
```
Wouldn't it be a more transparent solution to check if we are in a CAN FD mode of operation directly in the macro as soon as we can deduct the max bitrate from it ? (1Mbps for CAN and 5Mbps for CANFD)

henrikbrixandersen replied:
Some CAN controllers may impose other restrictions on the maximum bitrate than the maximum bitrates supported by the standard (e.g. some CAN-FD controllers only support 2Mbit/s). This is why I left it up to the individual driver to specify its maximum bitrate.

aaillet reviewed (commented):
Everything seems to work fine on H3ULCB board.
The standby pin management logic in the driver seems right to me but I'm not able to test it since the TCAN332GDCNT transceiver used on my board is not equipped with a standby gpio.
Too high bitrate guard is however working flawlessly on my setup.
H3ULCB dts also looks good to me.
However, here is some comments.
Thanks for this work.

henrikbrixandersen commented on include/drivers/can/transceiver.h:
```diff
@@ -0,0 +1,130 @@
+ * @param max maximum bitrate supported by the CAN controller
+ * @return the maximum bitrate supported by the CAN controller/transceiver combination
+ */
+#define CAN_TRANSCEIVER_DT_MAX_BITRATE_GET(node_id, max)				\
+	COND_CODE_1(DT_NODE_HAS_PROP(node_id, phys),					\
+		    MIN(DT_PROP(DT_PHANDLE(node_id, phys), max_bitrate), max),		\
+		    MIN(DT_PROP_OR(DT_CHILD(node_id, can_transceiver), max_bitrate, max), max))
```
The approach used here (as noted elsewhere in this PR) is compliant with upstream Linux devicetree bindings for CAN transceivers.

aaillet replied:
Ok with you about the can-transceiver dts but I'm struggling at finding an examples showing the use of `phys` or something close to it when declaring a standby gpio in a linux device tree, can you link me one ?

aaillet replied:
Perfect, thanks for the links !

henrikbrixandersen commented:
Rebased to fix merge conflict in `drivers/can/CMakeLists.txt` after e9c9caa80d16d3bd44590513ceb073b95c9ddb8e.

martinjaeger commented on boards/arm/rddrone_fmuk66/rddrone_fmuk66.dts:
```diff
@@ -57,6 +57,20 @@
 		};
 	};
+
+	tja1042_0: can-transceiver0 {
```
Intuitively, I would have chosen the naming in the opposite way: The specific manufacturer device type `tja1042` as the node name and the more generic `can-transceiver0` as the label. Maybe even (ab)use the address for numbering the different devices of same type.
```suggestion
	can-transceiver0: tja1042@0 {
```

This way the reference in the CAN controller does not change if the type / compatible of the transceiver changes.

Consider this just a remark. Definitely not blocking.

henrikbrixandersen replied:
Thanks. I have changed these to use the DTS nodelabel/name scheme applied in upstream Linux DTS. The devicetree specification does not (yet) mention CAN transceivers, but this naming scheme matches that used for USB and Ethernet PHYs.

martinjaeger commented on include/drivers/can.h:
```diff
@@ -766,8 +799,21 @@ static inline int can_set_bitrate(const struct device *dev,
+	uint32_t max_bitrate;
 	int ret;
 
+	ret = can_get_max_bitrate(dev, &max_bitrate);
+	if (ret == -ENOSYS) {
+		/* Maximum bitrate unknown */
+		max_bitrate = 0;
+	} else if (ret < 0) {
+		return ret;
+	}
+
+	if ((max_bitrate > 0) && (bitrate > max_bitrate)) {
+		return -ENOTSUP;
+	}
+
```
This function is getting pretty long for a `static inline`. See also recommendations in [Linux Kernel coding style guide](https://www.kernel.org/doc/html/latest/process/coding-style.html#the-inline-disease). Should we move it to `can_common.c`?

henrikbrixandersen replied:
Good idea. I have moved it to `can_common.c` in a separate commit to make reviewing of the changes easier.

henrikbrixandersen committed:
drivers: can: move can_set_bitrate() to can_common.c

Move the can_set_bitrate() function to can_common.c as it is getting
quite long for a static inline function.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen committed:
devicetree: add devicetree/can.h

This contains accessor macros for getting the maximum bitrate supported
by a CAN controller/transceiver combination.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen commented:
Rebased to address the following issues:
- Moved `can_set_bitrate()` implementation to `can_common.c` as advised by @martinjaeger 
- Renamed CAN transceiver DTS nodelabels/names as advised by @martinjaeger 
- Moved and renamed the CAN transceiver devicetree macros to `include/devicetree/can.h` (and added test cases) as advised by @mbolivar-nordic (via Discord)

martinjaeger commented on dts/bindings/can/can-controller.yaml:
```diff
@@ -30,3 +30,42 @@ properties:
+      description: |
+        Actively controlled CAN transceiver.
+
+        Example:
+          tja1040: can-transceiver {
+            compatible = "nxp,tja1040", "can-transceiver-gpio";
+            standby-gpios = <gpioa 0 GPIO_ACTIVE_HIGH>;
+            max-bitrate = <1000000>;
+            #phy-cells = <0>;
+          };
+
+          &can0 {
+            status = "okay";
+
+            phys = <&tja1040>;
+          };
```
Should be updated to match the code change as discussed before.

henrikbrixandersen replied:
Thanks, fixed.

henrikbrixandersen commented:
Rebased to update dts binding example as noted by @martinjaeger.

nandojve commented on boards/arm/sam_e70_xplained/sam_e70_xplained-common.dtsi:
```diff
@@ -157,4 +157,8 @@ zephyr_udc0: &usbhs {
 	pinctrl-0 = <&pb3a_can0_rx0 &pb2a_can0_tx0>;
 	bus-speed = <125000>;
 	bus-speed-data = <1000000>;
+
+	can-transceiver {
+		max-bitrate = <5000000>;
+	};
```
Hi @henrikbrixandersen ,
The sam_v71_xult board have a 5Mbit can-fd transceiver too.
Do you mind add can entry for sam_v71_xult board?

henrikbrixandersen replied:
@nandojve Added, thank you.

nandojve replied:
Thank you for this PR!

henrikbrixandersen committed:
boards: arm: sam_v71_xult: add CAN transceiver

The Atmel SAM V71 Xplained Ultra development board is equipped with a
Microchip ATA6561 CAN transceiver with a maximum bitrate of 5Mbit/s.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

mbolivar-nordic reviewed (dismissed):
Devicetree portions look great. You have my heartfelt thanks for a really easy review.

henrikbrixandersen commented:
@alexanderwachter ping

karstenkoenig reviewed (commented):
Cool! Just some nits on the mcp2551 transceiver on the olimex board

henrikbrixandersen commented on boards/arm/olimexino_stm32/olimexino_stm32.dts:
```diff
@@ -40,6 +40,13 @@
 
+	transceiver0: can-phy0 {
+		compatible = "microchip,mcp2551", "can-transceiver-gpio";
+		max-bitrate = <5000000>;
```
Fixed. Thank you!

henrikbrixandersen commented on boards/arm/olimexino_stm32/olimexino_stm32.dts:
```diff
@@ -40,6 +40,13 @@
+	transceiver0: can-phy0 {
+		compatible = "microchip,mcp2551", "can-transceiver-gpio";
+		max-bitrate = <5000000>;
+		standby-gpios = <&gpiod 21 GPIO_ACTIVE_HIGH>;
```
Oh, nice catch. Fixed, thanks.
